Better determine concurrency from host environment to use for JavaTestRunner

andrew-m-leonard opened the pull request:
JavaTestRunner can use too high a concurrency based purely on CPU count, if such a host has insufficient memory it can cause memory starvation and Jenkins channel failures.

This PR based off the logic used in openjdk/openjdk.mk, optimises the concurrency to min(CPUs,MemInGb)

Signed-off-by: Andrew Leonard <anleonar@redhat.com>

andrew-m-leonard commented:
**Test Grinders on various hosts:**

160 core 512Gb Aarch64 Linux:
```
11:32:38  ../jck.mk:147: Jck concurrency determined from host environment OS=Linux, CPUs=160, Memory=501 Gb. Using concurrency=160
```

Mesbah-Alam commented:
@andrew-m-leonard , Hi Andrew, thanks for this improvement. Could we please run a test on Windows to ensure all is well there?

andrew-m-leonard commented:
> @andrew-m-leonard , Hi Andrew, thanks for this improvement. Could we please run a test on Windows to ensure all is well there?

Thanks, just doing that now...

Mesbah-Alam commented:
Tested on Windows, looks fine: 

```
11:07:54  ../jck.mk:151: Jck concurrency determined from host environment OS=CYGWIN_NT-10.0, CPUs=8, Memory=15 Gb. Using concurrency=8
```

Mesbah-Alam reviewed (approved):
LGTM

andrew-m-leonard commented:
@Mesbah-Alam so interesting the memory size calc fails on Windows, which is odd as I copied it from openjdk.mk, then realized it doesn't work there anyway ! https://ci.adoptopenjdk.net/view/Test_openjdk/job/Test_openjdk11_hs_sanity.openjdk_x86-64_windows/697/consoleFull

andrew-m-leonard commented:
I'll try and fix both...

sxa reviewed (commented):
Can this still be overridden when you run the job in jenkins?
Have we seen a failure due to this on any machines other than the Skytap AIX ones, and have we looked to see if it's really an out of memory situation or is that just a guess at present? If we can't utilise all the cores we have we should consider adjusting the specification of the machines accordingly.

Could one process for every 512Mb of RAM be used instead (1Gb seems generous to me) this making a 32-core 8Gb system use concurrency=16 by default

(Also as an aside I'm a bit worried that we're duplicating quite a lot of logic between here and other places where such a detection occurs)

andrew-m-leonard commented:
@Mesbah-Alam should be fixed, also fixed openjdk.mk and added missing AIX logic there too:
openjdk tests:
Windows: https://ci.adoptopenjdk.net/view/Test_openjdk/job/Test_openjdk11_hs_sanity.openjdk_x86-64_windows/698/
AIX: https://ci.adoptopenjdk.net/view/Test_openjdk/job/Test_openjdk11_j9_sanity.openjdk_ppc64_aix/490/

Running jck test as well

andrew-m-leonard commented:
> Can this still be overridden when you run the job in jenkins? Have we seen a failure due to this on any machines other than the Skytap AIX ones, and have we looked to see if it's really an out of memory situation or is that just a guess at present? If we can't utilise all the cores we have we should consider adjusting the specification of the machines accordingly.
> 
> Could one process for every 512Mb of RAM be used instead (1Gb seems generous to me) this making a 32-core 8Gb system use concurrency=16 by default
> 
> (Also as an aside I'm a bit worried that we're duplicating quite a lot of logic between here and other places where such a detection occurs)

Yes, can be overridden

Suspicion yes, but i've done multiple tests with lower concurrency and failures have not occurred, so seems likely on AIX.

I have suspicion on other platforms too, Alpine one mentioned earlier.

andrew-m-leonard commented:
@sxa Some tests do need a lot of memory, some of the compiler tests especially.

Mesbah-Alam commented:
@andrew-m-leonard , Interestingly, it worked at my end on x64 Windows before you updated the PR, Grinder_JCK/979: 

```
11:07:52  ../jck.mk:36: Running under cygwin, looking for Windows perl on path
11:07:53  ../jck.mk:41: Found perl in /cygdrive/c/Strawberry/perl/bin/perl.exe
11:07:54  ../jck.mk:151: Jck concurrency determined from host environment OS=CYGWIN_NT-10.0, CPUs=8, Memory=15 Gb. Using concurrency=8
11:07:55  
11:07:55  ===============================================
11:07:55  Running test jck-compiler-api-javax_annotation_0 ...
11:07:56  ===============================================
11:07:57  jck-compiler-api-javax_annotation_0 Start Time: Fri Dec  9 08:07:21 2022 Epoch Time (ms): 1670602041491
11:07:57  variation: NoOptions
11:07:57  JVM_OPTIONS:  
``` 

Thanks for fixing it at your end!

sxa commented:
The reporting of 32 processors on AIX using the lcpu value is probably not a good one to use. The machine that threw up the problem we're looking to resolve here has 4 CPUs with 8 threads/CPU (SMT8) which shows as 32 `lcpu`s. While the optimal number is potentially somewhere between 4 and 32, I would suggest that perhaps we should only override this for AIX, and either use the real number of processors (Ideally plus one) or some other combination which takes into account the threading multiplier.

To get the number of CPUs on AIX use: `lparstat -i | awk '/Online Virtual CPUs/{print$NF}'`

We should probably also change this wherever the test harness decides on it it instead of exclusively for the JCK since the autodetected value is probably not optimal for highly threaded processors.

andrew-m-leonard commented:
@sxa it's worth bearing in mind the JTreg openjdk tests are even more generous:
https://github.com/adoptium/aqa-tests/blob/61596d279e5f443c524aa0688cb2ddf371c33921/openjdk/openjdk.mk#L47
So an 32cpu 8Gb machine would use concurrency=4 for openjdk jtreg tests.

I do think some of the intermittent jck problems are because the higher concurrency obviously consumes more resource, but also lines up the higher chance of some of the high memory/longer running individual tests coinciding with other high memory ones.

I'd like us to give back some more generous room, so as to see the affect, hopefully a positive one. If we just ease back a little, we will probably still see those issues, and dismiss this as the problem.

andrew-m-leonard commented:
> The reporting of 32 processors on AIX using the lcpu value is probably not a good one to use. The machine that threw up the problem we're looking to resolve here has 4 CPUs with 8 threads/CPU (SMT8) which shows as 32 `lcpu`s. While the optimal number is potentially somewhere between 4 and 32, I would suggest that perhaps we should only override this for AIX, and either use the real number of processors (Ideally plus one) or some other combination which takes into account the threading multiplier.
> 
> To get the number of CPUs on AIX use: `lparstat -i | awk '/Online Virtual CPUs/{print$NF}'`
> 
> We should probably also change this wherever the test harness decides on it it instead of exclusively for the JCK since the autodetected value is probably not optimal for highly threaded processors.

@sxa As you say those AIX nodes seem more than 4 CPU capable, I would hope the 32 lcpu value is representable of its capability, otherwise what's the point of that value?

andrew-m-leonard commented:
> The reporting of 32 processors on AIX using the lcpu value is probably not a good one to use. The machine that threw up the problem we're looking to resolve here has 4 CPUs with 8 threads/CPU (SMT8) which shows as 32 `lcpu`s. While the optimal number is potentially somewhere between 4 and 32, I would suggest that perhaps we should only override this for AIX, and either use the real number of processors (Ideally plus one) or some other combination which takes into account the threading multiplier.
> 
> To get the number of CPUs on AIX use: `lparstat -i | awk '/Online Virtual CPUs/{print$NF}'`
> 
> We should probably also change this wherever the test harness decides on it it instead of exclusively for the JCK since the autodetected value is probably not optimal for highly threaded processors.

I'm tempted to leave it base on the AIX lcpu, since it gives a concurrency of 8, which I think is about right ?

sxa commented:
> > We should probably also change this wherever the test harness decides on it it instead of exclusively for the JCK since the autodetected value is probably not optimal for highly threaded processors.
> @sxa As you say those AIX nodes seem more than 4 CPU capable, I would hope the 32 lcpu value is representable of its capability, otherwise what's the point of that value?

That's a long complex processor architecture question and it depends on the workload :-) I would be very surprised if there was any point in running 32 but in the interests of finding out:

Grinders 2183 through 2186 are running the sanity suite on the machine with concurrencies set to default (32), 16, 8 and 4 respectively.

andrew-m-leonard commented:
> Grinders 2183 through 2186 are running the sanity suite on the machine with concurrencies set to default (32), 16, 8 and 4 respectively.

@sxa the time variation might be interesting. I can only elude to which might be high memory tests, a number of the compiler tests use deep recursion scenarios, but they are mainly extended suite.

sxa commented:
> Grinders 2183 through 2186 are running the sanity suite on the machine with concurrencies set to default (32), 16, 8 and 4 respectively.

Those disappeared, so I'll post the results of another test with a couple of suites to give an idea of the difference caused by different concurrency values on a Skytap 4-core, SMT8 system:

concurrency | Execution time
--- | ---
35 (default) | 34m09
16 | 40m06
8 | 52m58
5 | 1h07
4 | 1h13

Dropping from a concurrency value of 35 to 8 would give a 50% increase in runtime under this scenario.

andrew-m-leonard commented:
> > Grinders 2183 through 2186 are running the sanity suite on the machine with concurrencies set to default (32), 16, 8 and 4 respectively.
> 
> Those disappeared, so I'll post the results of another test with a couple of suites to give an idea of the difference caused by different concurrency values on a Skytap 4-core, SMT8 system:
> concurrency 	Execution time
> 35 (default) 	34m09
> 16 	40m06
> 8 	52m58
> 5 	1h07
> 4 	1h13
> 
> Dropping from a concurrency value of 35 to 8 would give a 50% increase in runtime under this scenario.

Interesting thanks @sxa 
So i'd err on the side of not thrashing the machine at the risk of causing failures or crippling the machine or agent, so 16 or 8 would be my suggested target, probably 8 to try and ensure consistent Green runs

andrew-m-leonard commented:
@sxa going to keep log of expected process termination/errors here...

andrew-m-leonard commented:
https://ci.eclipse.org/temurin-compliance/job/Test_openjdk8_hs_extended.jck_ppc64_aix/51/console:
```
16:18:33  testsuite "/home/jenkins/jck_root/JCK8-unzipped/JCK-runtime-8c";
16:18:33  workDirectory -create -overwrite /home/jenkins/workspace/Test_openjdk8_hs_extended.jck_ppc64_aix/aqa-tests/TKG/output_16706762027040/jck-runtime-api-java_util_0/workdir;
16:18:33  tests api/java_util;
16:18:33  concurrency 33;
16:18:33  timeoutfactor 4;
16:18:33  keywords !interactive;
16:18:33  set jck.env.testPlatform.os "other";
16:18:33  set jck.env.testPlatform.fileSep "/";
16:18:33  set jck.env.testPlatform.pathSep ":";
16:18:33  set jck.env.runtime.testExecute.cmdAsString "/home/jenkins/workspace/Test_openjdk8_hs_extended.jck_ppc64_aix/openjdkbinary/j2sdk-image/bin/java";
16:18:33  set jck.env.runtime.testExecute.otherOpts "   -Djava.ext.dirs=/home/jenkins/jck_root/JCK8-unzipped/JCK-runtime-8c/lib/extensions:/home/jenkins/workspace/Test_openjdk8_hs_extended.jck_ppc64_aix/openjdkbinary/j2sdk-image/jre/lib/ext -Djava.security.properties=/home/jenkins/workspace/Test_openjdk8_hs_extended.jck_ppc64_aix/aqa-tests/TKG/output_16706762027040/jck-runtime-api-java_util_0/security.properties ";
16:18:33  set jck.excludeList.customFiles "/home/jenkins/jck_root/JCK8-unzipped/JCK-runtime-8c/lib/jck8c.jtx   /home/jenkins/jck_run/jdk8/aix/temurin.jtx  ";
16:18:33  runTests;
16:18:33  writeReport -type xml /home/jenkins/workspace/Test_openjdk8_hs_extended.jck_ppc64_aix/aqa-tests/TKG/output_16706762027040/jck-runtime-api-java_util_0/report;
16:18:33  <<<<<<<<
16:18:33  Starting sub-process: com.sun.javatest.agent.AgentMain
16:18:33  Full command: [/home/jenkins/workspace/Test_openjdk8_hs_extended.jck_ppc64_aix/openjdkbinary/j2sdk-image/bin/java, -Djavatest.security.allowPropertiesAccess=true, -Djava.security.policy=/home/jenkins/jck_root/JCK8-unzipped/JCK-runtime-8c/lib/jck.policy, -classpath, /home/jenkins/jck_root/JCK8-unzipped/JCK-runtime-8c/lib/javatest.jar:/home/jenkins/jck_root/JCK8-unzipped/JCK-runtime-8c/classes, com.sun.javatest.agent.AgentMain, -passive]
16:18:33  Starting sub-process: tnameserve
16:18:33  Full command: [/home/jenkins/workspace/Test_openjdk8_hs_extended.jck_ppc64_aix/openjdkbinary/j2sdk-image/bin/tnameserv, -ORBInitialPort, 9876]
16:18:33  Running JCK in multijvm way with Agent off
16:18:33  Starting sub-process: jck
16:18:33  Full command: [/home/jenkins/workspace/Test_openjdk8_hs_extended.jck_ppc64_aix/openjdkbinary/j2sdk-image/bin/java, -jar, /home/jenkins/jck_root/JCK8-unzipped/JCK-runtime-8c/lib/javatest.jar, -config, /home/jenkins/workspace/Test_openjdk8_hs_extended.jck_ppc64_aix/aqa-tests/TKG/../jck/jtrunner/config/jck8c/runtime.jti,  @/home/jenkins/workspace/Test_openjdk8_hs_extended.jck_ppc64_aix/aqa-tests/TKG/output_16706762027040/jck-runtime-api-java_util_0/JTB/generated.jtb]
16:18:33  Initial Naming Context:
16:18:33  IOR:000000000000002b49444c3a6f6d672e6f72672f436f734e616d696e672f4e616d696e67436f6e746578744578743a312e30000000000001000000000000009a000102000000000e32302e36312e3133362e32313300269400000045afabcb0000000020000f424000000001000000000000000200000008526f6f74504f41000000000d544e616d65536572766963650000000000000008000000010000000114000000000000020000000100000020000000000001000100000002050100010001002000010109000000010001010000000026000000020002
16:18:33  TransientNameServer: setting port for initial object references to: 9876
16:18:33  Ready.
16:18:36  Dec 10, 2022 4:18:35 PM Harness starting test run with configuration "jck_runtime"...
16:20:44  gmake[1]: *** Waiting for unfinished jobs....
16:20:44  gmake[2]: *** Waiting for unfinished jobs....
16:20:44  gmake[4]: *** Waiting for unfinished jobs....
16:20:44  gmake[3]: *** Waiting for unfinigmake: *shed job** Waitis....
16:20:44  ng for u[thread 515 also had an error]
16:20:44  [thread 1543 also had an error]
16:20:44  [thread 772 also had an error]
16:20:44  nfinished jobs....
16:20:44  #
16:20:44  # A fatal error has been detected by the Java Runtime Environment:
16:20:44  #
16:20:44  #  SIGDANGER (0x21) at pc=0x090000000052c8c4, pid=5243026, tid=0x0000000000000203
16:20:44  #
16:20:44  # JRE version: OpenJDK Runtime Environment (8.0_362-b05) (build 1.8.0_362-b05)
16:20:44  # Java VM: OpenJDK 64-Bit Server VM (25.362-b05 mixed mode aix-ppc64 compressed oops)
16:20:44  # Problematic frame:
16:20:44  # C  [libpthread.a+0xffffffff]  _event_sleep+0x584
16:20:44  #
16:20:44  # Core dump written. Default location: /home/jenkins/workspace/Test_openjdk8_hs_extended.jck_ppc64_aix/aqa-tests/TKG/output_16706762027040/jck-runtime-api-java_util_0/core or core.5243026
16:20:44  #
16:20:44  #
16:20:44  # A fatal error has been detected by the Java Runtime Environment:
16:20:44  #
16:20:44  #  SIGDANGER (0x21) at pc=0x0900000000530a44, pid=10813640, tid=0x0000000000000304
16:20:44  #
16:20:44  # JRE version: OpenJDK Runtime Environment (8.0_362-b05) (build 1.8.0_362-b05)
16:20:44  # Java VM: OpenJDK 64-Bit Server VM (25.362-b05 mixed mode aix-ppc64 compressed oops)
16:20:44  # Problematic frame:
16:20:44  # C  [libpthread.a+0xffffffff]  _sigsetmask+0xe4
16:20:44  #
16:20:44  # Core dump written. Default location: /home/jenkins/workspace/Test_openjdk8_hs_extended.jck_ppc64_aix/aqa-tests/TKG/output_16706762027040/jck-runtime-api-java_util_0/core or core.10813640
16:20:44  #
16:20:44  #
16:20:44  # A fatal error has been detected by the Java Runtime Environment:
16:20:44  #
16:20:44  #  SIGDANGER (0x21) at pc=0x090000000052c8c4, pid=17563792, tid=0x0000000000000203
16:20:44  #
16:20:44  # JRE version: OpenJDK Runtime Environment (8.0_362-b05) (build 1.8.0_362-b05)
16:20:44  # Java VM: OpenJDK 64-Bit Server VM (25.362-b05 mixed mode aix-ppc64 compressed oops)
16:20:44  # Problematic frame:
16:20:44  # C  [libpthread.a+0xffffffff]  _event_sleep+0x584
16:20:44  #
16:20:44  # Core dump written. Default location: /home/jenkins/workspace/Test_openjdk8_hs_extended.jck_ppc64_aix/aqa-tests/TKG/output_16706762027040/jck-runtime-api-java_util_0/core or core.17563792
16:20:44  #
16:20:44  # An error report file with more information is saved as:
16:20:44  # /home/jenkins/workspace/Test_openjdk8_hs_extended.jck_ppc64_aix/aqa-tests/TKG/output_16706762027040/jck-runtime-api-java_util_0/hs_err_pid17563792.log
16:20:44  #
16:20:44  # A fatal error has been detected by the Java Runtime Environment:
16:20:44  #
16:20:44  #  SIGDANGER (0x21) at pc=0x090000000052c8c4, pid=21692638, tid=0x0000000000000203
16:20:44  #
16:20:44  # JRE version: OpenJDK Runtime Environment (8.0_362-b05) (build 1.8.0_362-b05)
16:20:44  # Java VM: OpenJDK 64-Bit Server VM (25.362-b05 mixed mode aix-ppc64 compressed oops)
16:20:44  # Problematic frame:
16:20:44  # C  [libpthread.a+0xffffffff]  _event_sleep+0x584
16:20:44  #
16:20:44  # Core dump written. Default location: /home/jenkins/workspace/Test_openjdk8_hs_extended.jck_ppc64_aix/aqa-tests/TKG/output_16706762027040/jck-runtime-api-java_util_0/core or core.21692638
16:20:44  #
```

andrew-m-leonard commented:
```
23:30:43  testsuite "/home/jenkins/jck_root/JCK11-unzipped/JCK-compiler-11";
23:30:43  workDirectory -create -overwrite /home/jenkins/workspace/Test_openjdk11_hs_extended.jck_x86-64_alpine-linux/aqa-tests/TKG/output_16706959873197/jck-compiler-lang-EXPR-group3_0/workdir;
23:30:43  tests lang/EXPR/expr138 lang/EXPR/expr139 lang/EXPR/expr140 lang/EXPR/expr142 lang/EXPR/expr143 lang/EXPR/expr144 lang/EXPR/expr146 lang/EXPR/expr148 lang/EXPR/expr149 lang/EXPR/expr150 lang/EXPR/expr151 lang/EXPR/expr152 lang/EXPR/expr153 lang/EXPR/expr154 lang/EXPR/expr155 lang/EXPR/expr156 lang/EXPR/expr157 lang/EXPR/expr158 lang/EXPR/expr159 lang/EXPR/expr160 lang/EXPR/expr161 lang/EXPR/expr162 lang/EXPR/expr163 lang/EXPR/expr164 lang/EXPR/expr165 lang/EXPR/expr166 lang/EXPR/expr167 lang/EXPR/expr168 lang/EXPR/expr169 lang/EXPR/expr170 lang/EXPR/expr171 lang/EXPR/expr172 lang/EXPR/expr173 lang/EXPR/expr174 lang/EXPR/expr175 lang/EXPR/expr176 lang/EXPR/expr177 lang/EXPR/expr178 lang/EXPR/expr179 lang/EXPR/expr180 lang/EXPR/expr181 lang/EXPR/expr182 lang/EXPR/expr183 lang/EXPR/expr184 lang/EXPR/expr185 lang/EXPR/expr186 lang/EXPR/expr187 lang/EXPR/expr188 lang/EXPR/expr189 lang/EXPR/expr190 lang/EXPR/expr191 lang/EXPR/expr192 lang/EXPR/expr193 lang/EXPR/expr194 lang/EXPR/expr195 lang/EXPR/expr196 lang/EXPR/expr197 lang/EXPR/expr198 lang/EXPR/expr199 lang/EXPR/expr200 lang/EXPR/expr201 lang/EXPR/expr202 lang/EXPR/expr203 lang/EXPR/expr204 lang/EXPR/expr205 lang/EXPR/expr206;
23:30:43  concurrency 3;
23:30:43  timeoutfactor 4;
23:30:43  keywords compiler;
23:30:43  set jck.env.testPlatform.os "other";
23:30:43  set jck.env.testPlatform.fileSep "/";
23:30:43  set jck.env.testPlatform.pathSep ":";
23:30:43  set jck.env.compiler.testCompile.testCompileAPImultiJVM.cmdAsString "/home/jenkins/workspace/Test_openjdk11_hs_extended.jck_x86-64_alpine-linux/openjdkbinary/j2sdk-image/bin/java";
23:30:43  set jck.env.compiler.testCompile.otherOpts "-source 11 ";
23:30:43  set jck.env.compiler.compRefExecute.cmdAsString "/home/jenkins/workspace/Test_openjdk11_hs_extended.jck_x86-64_alpine-linux/openjdkbinary/j2sdk-image/bin/java";
23:30:43  set jck.env.compiler.compRefExecute.otherOpts "   ";
23:30:43  set jck.excludeList.customFiles "/home/jenkins/jck_root/JCK11-unzipped/JCK-compiler-11/lib/jck11.jtx   /home/jenkins/jck_run/jdk11/linux/temurin.jtx  ";
23:30:43  runTests;
23:30:43  writeReport -type xml /home/jenkins/workspace/Test_openjdk11_hs_extended.jck_x86-64_alpine-linux/aqa-tests/TKG/output_16706959873197/jck-compiler-lang-EXPR-group3_0/report;
23:30:43  <<<<<<<<
23:30:43  Running JCK in multijvm way with Agent off
23:30:43  Starting sub-process: jck
23:30:43  Full command: [/home/jenkins/workspace/Test_openjdk11_hs_extended.jck_x86-64_alpine-linux/openjdkbinary/j2sdk-image/bin/java, -jar, /home/jenkins/jck_root/JCK11-unzipped/JCK-compiler-11/lib/javatest.jar, -config, /home/jenkins/workspace/Test_openjdk11_hs_extended.jck_x86-64_alpine-linux/aqa-tests/TKG/../jck/jtrunner/config/jck11/compiler.jti,  @/home/jenkins/workspace/Test_openjdk11_hs_extended.jck_x86-64_alpine-linux/aqa-tests/TKG/output_16706959873197/jck-compiler-lang-EXPR-group3_0/JTB/generated.jtb]
23:30:47  Dec 10, 2022, 11:30:46 PM Harness starting test run with configuration "jck_compiler"...
23:33:18  make[4]: *** [autoGen.mk:425: jck-compiler-lang-EXPR-group3_0] Terminated
23:33:18  make[3]: *** [/home/jenkins/workspace/Test_openjdk11_hs_extended.jck_x86-64_alpine-linux/aqa-tests/TKG/../TKG/settings.mk:356: extended.jck-compiler.lang] Terminated
23:33:18  make[2]: *** [/home/jenkins/workspace/Test_openjdk11_hs_extended.jck_x86-64_alpine-linux/aqa-tests/TKG/../TKG/settings.mk:356: extended.jck-jck] Terminated
23:33:18  make[1]: *** [settings.mk:356: extended.jck-..] Terminated
23:33:18  make: *** [makefile:50: _extended.jck] Terminated
23:33:18  Terminated
```

sxa reviewed (changes requested):
Putting a block on this to avoid anyone merging until we've verified whether an upgrade to JDK11 as the agent machine resolves some of the errors we've seen, which would render this irrelevant.

smlambert commented:
Do we want/need this PR, or close it?

andrew-m-leonard commented:
JTRunner currently only uses CPU number as the basis of concurrency, we are happy that given a correctly configured infrastructure that using just that works.
